File: packages/api-models-py/skyportal_py_models/observing_runs.py
"""Request and response models for ``/api/observing_run``."""

from __future__ import annotations

import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.assignments import Assignment
from skyportal_py_models.groups import Group
from skyportal_py_models.instruments import Instrument
from skyportal_py_models.telescopes import Ephemeris
from skyportal_py_models.users import User


class ObservingRun(BaseModel):
    """A classical observing run (upstream ``ObservingRun``).

    ``sources`` stays ``dict`` because typing its entries as
    :class:`skyportal_py_models.sources.Source` would create an import cycle.
    The list endpoint returns ``to_dict()`` output (columns plus the
    eager-loaded ``instrument``); the single-run endpoint returns a
    hand-built dict that swaps ``created_at``/``modified``/``run_end_utc``
    for ``ephemeris`` and the run's ``assignments``.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    instrument_id: int | None = None
    calendar_date: datetime.date | None = None
    run_end_utc: datetime.datetime | None = None
    pi: str | None = None
    observers: str | None = None
    duration: int | None = None
    group_id: int | None = None
    owner_id: int | None = None
    ephemeris: Ephemeris | None = None
    instrument: Instrument | None = None
    group: Group | None = None
    owner: User | None = None
    assignments: list[Assignment] = Field(default_factory=list)
    sources: list[dict[str, Any]] = Field(default_factory=list)


class ObservingRunPost(BaseModel):
    """Payload for creating an observing run."""

    model_config = ConfigDict(extra="forbid")

    instrument_id: int
    calendar_date: str
    pi: str | None = None
    observers: str | None = None
    duration: int | None = None
    group_id: int | None = None


class ObservingRunPostResponse(BaseModel):
    """Result of creating an observing run."""

    model_config = ConfigDict(extra="forbid")

    id: int


class ObservingRunUpdate(BaseModel):
    """Payload for updating an observing run; every field is optional."""

    model_config = ConfigDict(extra="forbid")

    instrument_id: int | None = None
    calendar_date: str | None = None
    pi: str | None = None
    observers: str | None = None
    duration: int | None = None
    group_id: int | None = None

File: packages/api-models-ts/src/ObservingRuns.ts
/**
 * Request and response models for `/api/observing_run`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Assignments from "./Assignments.ts";
import * as Groups from "./Groups.ts";
import * as Instruments from "./Instruments.ts";
import * as Schemas from "./Schemas.ts";
import * as Telescopes from "./Telescopes.ts";
import * as Users from "./Users.ts";

/**
 * A classical observing run (upstream `ObservingRun`).
 *
 * `sources` stays free-form because typing its entries as
 * {@link skyportal-js/Sources!Source} would create an import cycle. The list
 * endpoint returns `to_dict()` output (columns plus the eager-loaded
 * `instrument`); the single-run endpoint returns a hand-built dict that swaps
 * `created_at`/`modified`/`run_end_utc` for `ephemeris` and the run's
 * `assignments`.
 *
 * @since 1.0.0
 * @category Models
 */
export const ObservingRun = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        instrument_id: Schemas.NullishInteger,
        calendar_date: Schemas.NullishString,
        run_end_utc: Schemas.NullishTimestamp,
        pi: Schemas.NullishString,
        observers: Schemas.NullishString,
        duration: Schemas.NullishInteger,
        group_id: Schemas.NullishInteger,
        owner_id: Schemas.NullishInteger,
        ephemeris: Schemas.nullish(Telescopes.Ephemeris),
        instrument: Schemas.nullish(Instruments.Instrument),
        group: Schemas.nullish(Groups.Group),
        owner: Schemas.nullish(Users.User),
        assignments: Schemas.list(Assignments.Assignment),
        sources: Schemas.list(Schemas.JsonObject),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type ObservingRun = v.InferOutput<typeof ObservingRun>;

/**
 * Payload for creating an observing run.
 *
 * `calendar_date` is the local calendar date of the run in ISO format, e.g.
 * `"2026-09-01"`; `duration` is the number of nights.
 *
 * @since 1.0.0
 * @category Models
 */
export interface ObservingRunPost {
    readonly instrument_id: number;
    readonly calendar_date: string;
    readonly pi?: string | undefined;
    readonly observers?: string | undefined;
    readonly duration?: number | undefined;
    readonly group_id?: number | undefined;
}

/**
 * Payload for updating an observing run; every field is optional.
 *
 * @since 1.0.0
 * @category Models
 */
export interface ObservingRunUpdate {
    readonly instrument_id?: number | undefined;
    readonly calendar_date?: string | undefined;
    readonly pi?: string | undefined;
    readonly observers?: string | undefined;
    readonly duration?: number | undefined;
    readonly group_id?: number | undefined;
}

/**
 * Result of creating an observing run.
 *
 * @since 1.0.0
 * @category Models
 */
export const ObservingRunPostResponse = Schemas.model(
    v.strictObject({ id: Schemas.Integer })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type ObservingRunPostResponse = v.InferOutput<typeof ObservingRunPostResponse>;
